### DISCUSSION TITLE
Mouse from ZMK not working

### DISCUSSION THREAD

**Discussion by Boggin** [urob/zmk-config#64] [Q&A]
I've a fork that I hadn't updated in a year: https://github.com/Boggin/zmk-config. 

As part of getting up to the latest Zephyr 3.5 I changed to using the ZMK mouse. I've got something wrong, however, as my mouse layer won't work. I can activate the layer but I can't move, scroll or click. Instead the keys output as per the base layer. 

I've tried checking my code against the main repo but I can't see any differences and everything else is working fine.

I've tried disconnecting from the laptop, removing the device, clearing all the bindings in the keyboard, adding the keeb back again. No joy.

Has anyone any other suggestions how I might debug my problem with using the mouse layer, please?

https://github.com/Boggin/zmk-config/blob/main/config/base.keymap#L190
https://github.com/Boggin/zmk-config/blob/main/config/combos.dtsi#L25
https://github.com/Boggin/zmk-config/blob/main/config/mouse.dtsi

**Comment by urob** [comment]
The mouse PR is still WIP and periodically requires updating the config to its newest api. If something breaks, it's worth looking closely at [`mouse.dtsi`](https://github.com/urob/zmk-config/blob/main/config/mouse.dtsi), the [`.config` files](https://github.com/urob/zmk-config/blob/main/config/planck_rev6.conf), and going through the [commit history](https://github.com/urob/zmk-config/commits/main/) to see if anything stands out (e.g., https://github.com/urob/zmk-config/commit/6043e4b27fed839383884e8ac83774c8c572bb17).

Other than that, some changes require [refreshing the HID descriptor](https://deploy-preview-2027--zmk.netlify.app/docs/features/bluetooth#refreshing-the-hid-descriptor), which is often the culprit if nothing else works.

8/1/2024: Edited for clarity

**Comment by Liam-Weitzel** [comment]
I had a similar issue on my linux laptop, refreshing the HID descriptor had no effect. When testing on my windows pc however, mouse movement & keys worked like a charm. Before the upgrade when using urobs ZMK [mouse-3.2](https://github.com/urob/zmk/tree/mouse-3.2) branch mouse movement & keys worked on both my linux and windows pcs.

I have tried flashing the previously working version again and for some reason the mouse movement & keys are still not working on my linux  pc.

**Reply by urob** [reply]
I currently don't have Linux set up outside of WSL so I can't really help here. If you haven't figured it out yet, it might be best to ask on the ZMK discord if others are having issues with Linux. From what you say, my first guess, though, would be that it's also a driver issue and it might be worth trying your distro's analog to "forget bluetooth device" and then repairing.

**Reply by joao-lobao** [reply]
Hi @Liam-Weitzel ! Any update on this issue? I've just got caught with this issue.

**Reply by Liam-Weitzel** [reply]
Hi @joao-lobao. I haven't found a solution but I did notice that when using a usb connection, the mouse movement and keys work great. Please keep me in the loop on any fix you find!! Ofcourse I will do the same :)

**Reply by joao-lobao** [reply]
@Liam-Weitzel thanks, I still haven't found out a solution too. From what I understand you've just upgraded recently from the 3.2 branch but I've been using successfully the most recent version for some months now. But somehow somewhere there is some recent problematic change because now my new builds even with my previously working commits introduce this problem. For now I have downgraded to urobs mouse-3.2. Will keep you updated as well. Thanks.

**Reply by urob** [reply]
This is to be expected due to the changes in the HID descriptor between the old mouse-3.2 PR and the updated work in [#2027](https://github.com/zmkfirmware/zmk/pull/2027). The issue is that most OS will cache the descriptor for Bluetooth devices, preventing the updated driver to work correctly.

If mouse works via cable but doesn't work via Bluetooth, chances are high that this is indeed the issue. Another indicator would be to try pairing with a new device that you have never paired with before. If in this case mouse works via Bluetooth: Bingo.

To resolve the issue, you'll need to [clear the Bluetooth cache of your operating system](https://deploy-preview-2027--zmk.netlify.app/docs/features/bluetooth#refreshing-the-hid-descriptor). On Windows, this can be done by (1) unpairing the keyboard from the computer, and then (2) selecting "forget the Bluetooth device". On Linux this setting will vary with your Distro and Desktop manager.

**Reply by joao-lobao** [reply]
Hey @urob, thanks. I've been using [main](https://github.com/urob/zmk) since the upgrade, at least for me that shouldn't be the problem. Only 4 days ago I've noted this. I've done it all: reseted the keyboard, cleared Bluetooth cache, tried other devices (not Windows though) and also via cable. But still none of the mouse functions work :(

**Reply by urob** [reply]
Agreed, this sounds like a different issue. Might be worth asking in the pointing-device channel of the ZMK Discord if anyone else on Linux experiences similar issues. (If you have a chance to test on Windows, this might also be useful for narrowing down the issue.)

**Reply by urob** [reply]
Oh, and just to be sure, there were some changes in the mouse API about 2 months ago (which were pulled into the branch that you are using). One now must explicitly load the `mouse_keys.dtsi` header (in addition to `mouse.h`). See my [comment above](https://github.com/urob/zmk-config/discussions/64#discussioncomment-9948804) for details.

**Reply by urob** [reply]
Nvm, just realize that it was you who had raised #52 in the first place :) Though, looking at [your config](https://github.com/joao-lobao/zmk-config/blob/master/config/redox.keymap) I don't see the `mouse_keys.dtsi` header?

**Reply by joao-lobao** [reply]
Yes, you don't see it because I've just pushed the temporary fix for the mouse-3.2 version. I'll try to get help on Discord and update if anything comes up. Thanks.

**Reply by urob** [reply]
I just realized I haven't kept my config in sync with my firmware repo! Try https://github.com/urob/zmk-config/commit/6043e4b27fed839383884e8ac83774c8c572bb17

**Reply by joao-lobao** [reply]
Yes, yesterday I found out that I needed the CONFIG_ZMK_MOUSE=y which must be mandatory now. But the issue now is that it only works via cable, but not working with BT even after profile clearing, keyboard reset, BT cache clear and trying other devices (still not Windows though).
I've went through the ZMK docs cause I'm still somewhat new to ZMK and tried a bunch of other system configs but with no success. Have any idea if something else is missing?
